sky/provision/fluidstack/instance.py

```diff
@@ -0,0 +1,220 @@
+        if ready_instance_cnt == config.count:
+            break
+        failed_instances = _filter_instances(
+            cluster_name_on_cloud,
+            ['timeout error', 'failed to create', 'out of stock'])
+        if failed_instances:
+            logger.error(f'Failed to create {len(failed_instances)}'
+                         f'instances for cluster {cluster_name_on_cloud}')
+            raise RuntimeError(
+                f'Failed to create {len(failed_instances)} instances.')
+
+        time.sleep(POLL_INTERVAL)
+    assert head_instance_id is not None, 'head_instance_id should not be None'
+    return common.ProvisionRecord(provider_name='fluidstack',
+                                  cluster_name=cluster_name_on_cloud,
+                                  region=region,
+                                  zone=None,
+                                  head_instance_id=head_instance_id,
+                                  resumed_instance_ids=[],
+                                  created_instance_ids=created_instance_ids)
+
+
+def wait_instances(region: str, cluster_name_on_cloud: str,
+                   state: Optional[status_lib.ClusterStatus]) -> None:
+    del region, cluster_name_on_cloud, state
+
+
+def stop_instances(
+    cluster_name_on_cloud: str,
+    provider_config: Optional[Dict[str, Any]] = None,
+    worker_only: bool = False,
+) -> None:
+    raise NotImplementedError()
+
+
+def terminate_instances(
+    cluster_name_on_cloud: str,
+    provider_config: Optional[Dict[str, Any]] = None,
+    worker_only: bool = False,
+) -> None:
+    """See sky/provision/__init__.py"""
+    del provider_config  # unused
+    instances = _filter_instances(cluster_name_on_cloud, None)
+    for inst_id, inst in instances.items():
+        logger.info(f'Terminating instance {inst_id}: {inst}')
+        if worker_only and inst['hostname'].endswith('-head'):
+            continue
+        try:
+            utils.FluidstackClient().delete(inst_id)
+        except Exception as e:  # pylint: disable=broad-except
+            with ux_utils.print_exception_no_traceback():
+                raise RuntimeError(
+                    f'Failed to terminate instance {inst_id}: '
+                    f'{common_utils.format_exception(e, use_bracket=False)}'
+                ) from e
+
+
+def get_cluster_info(
+        region: str,
+        cluster_name_on_cloud: str,
+        provider_config: Optional[Dict[str, Any]] = None) -> common.ClusterInfo:
+    del region, provider_config  # unused
+    running_instances = _filter_instances(cluster_name_on_cloud, ['running'])
+    instances: Dict[str, List[common.InstanceInfo]] = {}
+    head_instance_id = None
+    for instance_id, instance_info in running_instances.items():
+        instance_id = instance_info['id']
+        instances[instance_id] = [
+            common.InstanceInfo(
+                instance_id=instance_id,
+                internal_ip=instance_info['ip_address'],
```

Comment:
We need the private IP to have it work for job scheduling on multi-node cluster. Is it possible to set the private IP for the internal IP instead?

Reply:
If the `instance_info` does not return the private IP of the instance, a easy way to solve this is to `ssh` into the instance and get the private IP manually.

```python
from sky import authentication as auth
_GET_INTERNAL_IP_CMD = 'ip -4 -br addr show | grep UP | grep -Eo "(10\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)|172\.(1[6-9]|2[0-9]|3[0-1]))\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)"'

runners = command_runner.SSHCommandRunner.make_runner_list(
    [info['ip_address'] for info in running_instances],
    ssh_user=running_instances[0]['default_user_name']
    ssh_private_key=auth.PRIVATE_SSH_KEY_PATH)

def get_internal_ip(node_info: Dict[str, Any]) -> str:
    runner = command_runner.SSHCommandRunner(node_info['ip_address'], ssh_user=node_info['default_user_name'], ssh_private_key=auth.PRIVATE_SSH_KEY_PATH)
    rc, stdout, stderr = runner.run(_GET_INTERNAL_IP_CMD, require_outputs=True, stream_logs=False)
    subprocess_utils.handle_returncode(rc, _GET_INTERNAL_IP_CMD, 'Failed get obtain private IP from node', stderr=stdout+stderr)
    node_info['internal_ip'] = stdout.strip()

subprocess.run_in_parallel(get_internal_ips, running_instances)
for instance_id, instance_info in running_instances.items():
    instance_id = instance_info['id']
    instances[instance_id] = [
        common.InstanceInfo(
            instance_id=instance_id,
            internal_ip=instance_info['internal_ip'],
            external_ip=instance_info['ip_address'],
            ssh_port=instance_info['ssh_port'],
            tags={},
        )
    ]
```

Reply:
Just a kind bump for adopting the changes from the commit https://github.com/skypilot-org/skypilot/commit/781c630e69f1af517df340b0185472098cc77fea, otherwise the multi-node support will still fail.